course5/homework06/main.cc

```diff
@@ -0,0 +1,266 @@
+    assert(strncmp_result == 0);
+
+    for (auto i = 0; i < expected_length; i++)
+    {
+        assert(s[i] == expected_content[i]);
+    }
+
+    std::cout << "Passed: " << test_name << std::endl;
+}
+
+template<typename StrT, typename StrT2>
+void testEquality(const char* test_name, const StrT& s1, const StrT2& s2, bool are_equal)
+{
+    if (are_equal)
+    {
+        assert(s1 == s2);
+        assert(s2 == s1);
+    }
+    else
+    {
+        assert(s1 != s2);
+        assert(s2 != s1);
+    }
+
+    std::cout << "Passed: " << test_name << std::endl;
+}
+
+int main(int, char**)
+{
+    using namespace simple_string;
+    using namespace std::literals;
+
+    /* Basic tests for construction and assignment */
+    String defaultS;
+    String sFromChar = "word";
+    String copyConstructed = sFromChar;
+
+    String toBeAssigned = "movetest";
+    String moveConstructed = std::move(toBeAssigned);
+
+    toBeAssigned = "assignment";
+    String copyAssigner = "CopyAssignER";
+    String moveAssigner = "MoveAssignER";
+    String copyAssigned;
+    copyAssigned = copyAssigner;
+    String moveAssigned;
+    moveAssigned = std::move(moveAssigner);
+
+    copyAssigner[2] = 'z';
+
+    const int num_test_cases = 8;
+    std::array<const char *, num_test_cases> test_names = {
+        {
+            "default constructor",
+
+            "const char * construction",
+            "copy construction",
+            "move construction",
+
+            "const char * assignment",
+            "copy assignment",
+            "move assignment",
+
+            "[] assignment"
+        }
+    };
+    std::array<String, num_test_cases> inputs = {
+        {
+            defaultS,
+            sFromChar, copyConstructed, moveConstructed,
+            toBeAssigned, copyAssigned, moveAssigned,
+            copyAssigner
+        }
+    };
+    std::array<const char *, num_test_cases> expected_contents = {
+        {
+            "",
+            "word", "word", "movetest",
+            "assignment", "CopyAssignER", "MoveAssignER",
+            "CozyAssignER",
+        }
+    };
+
+    for (auto i = 0; i < num_test_cases; i++)
+    {
+        test(test_names[i], inputs[i], expected_contents[i]);
+    }
+
+    /* Tests for concatenation */
+    String concatStr = "Hello ";
+    String concatToConstChar = concatStr + "World!";
+    test("concat initialization w const char *", concatToConstChar, "Hello World!");
+
+    String concatStr2 = "World Again!";
+    String concatToString = concatStr + concatStr2;
+    test("concat initialization w String", concatToString, "Hello World Again!");
+
+    concatToString += " And again.";
+    test("+= w const char *", concatToString, "Hello World Again! And again.");
+
+    String concatStr3 = " And again twice.";
+    concatToString += concatStr3;
+    test("+= w String", concatToString, "Hello World Again! And again. And again twice.");
+
+    /* Test out of bounds error with indexing */
```

Comment:
Yes, but only if there is a ```catch``` block. Otherwise the program will just crash.
But if you will catch your exception, then there should be no leak.